### ISSUE TITLE
Group consecutive elements in a stream dependending on a common value

### ISSUE THREAD

**Issue by loicdescotte** [zio/zio#2415]
Hi,

I've found the need to group consecutive elements of a stream, that share a common value

TLDR : It is possible to do it with mapAccum and we could provide an "out of the box" method.

Example :  

```scala

case class Element(groupId: Int, value: String)

 val stream = Stream(
    Element(1, "1a"), Element(1, "1b"), Element(1, "1c"), Element(1, "1d"), Element(1, "1e"),
    Element(2, "2a"), Element(2, "2b"), Element(2, "2c"), Element(2, "2d"), Element(2, "2e"),
    Element(3, "3a"), Element(3, "3b"), Element(3, "3c"), Element(3, "3d"), Element(3, "3e"),
    Element(1, "1a"), Element(1, "1b"), Element(1, "1c"), Element(1, "1d"), Element(1, "1e")
  )
```
 
I expect to have 4 sub-streams as (groupId = 1, groupId = 2, groupId = 3, groupId = 1).

The last one is using groupId= 1 again to show that it's not a groupBy and I need to be able to start processing elements (including the ones from the first groupId = 1 group) before the end of the stream.

With Akka Streams it can be done with a sliding and `Source.splitAt`.

With ZIO  @iravid suggested me to use mapAccum like this :

```scala
stream.mapAccum(None: Option[(Int, List[Element])]) { (accum, element) =>
    accum match {
      case None => (Some((element.groupId, List(element))), None)
      case Some((groupId, elements)) if groupId == element.groupId => (Some((element.groupId, element :: elements)), None)
      case Some((_, elements)) => (Some(element.groupId, List(element)), Some(elements))
    }
  }.unNone.runCollect
```

It works well.

If you think this problem is common enough, the idea would be to provide a method to consecutive group elements that share a common attribute or a satisfy a common predicate. I don't have a good name for that yet :/

FYI, The Akka Streams solution with sliding + splitAt : 

```scala
source.prepend(Source.single(Element(0,"")))
    .sliding(2) // add previous element
    .splitWhen(e => e(0).groupId != 0 && e(0).groupId != e(1).groupId)
    .map{ case Vector(_, current) => current} //remove previous
    .fold(Seq[String]())(_ :+ _.value)
    .mergeSubstreams
    .runWith(Sink.foreach(println))
```

**Comment by loicdescotte** [comment]
Github pull/push is blocked on the network of the company where I'm working currenlty but I could make a PR with the mapAccum solutions in a few days if it's ok for everybody.

**Comment by atamurius** [comment]
Wouldn't `mapAccum + unNone` loose last group as there would be no element for last `case` branch?

**Comment by jypma** [comment]
I'd be very interested in a more general equivalent to akka's `SubFlow` type (what `splitWhen`) returns.

Not sure if an actual `SubZStream` would be necessary on ZIO. Perhaps a signature like
```scala
def splitWhen(p: T => Boolean): ZStream[R,E, ZStream[R,E,T]] = {
 ???
}
```
could just work, by having something keep track of which sub-stream is currently pulling, closing and opening sub-stream elements as the predicate is hit...

One could then build the above accumulating behavior by folding the sub-streams in the desired fashion.

**Comment by hearnadam** [comment]
@loicdescotte do you still need this? Feel free to open a PR.

**Comment by loicdescotte** [comment]
Hi, I don't need it personally but it seemed to be a useful feature at the time
